### ISSUE TITLE
Ray tune on Mac M2/M1 never stop

### ISSUE THREAD

**Issue by ngzk-tsubasa** [ray-project/ray#45797]
### What happened + What you expected to happen

Hi, 
I’m not sure if this is a bug. If I posted in the wrong place, I will move it to the appropriate forum. Sorry for any inconvenience.

I have set up a Ray cluster with the following configuration:
- Head node: Ubuntu with 12 CPUs and 1 GPU
- Worker nodes: One Ubuntu with 24 CPUs, one Mac Studio M2, and one Mac Mini M1.
And setup following: https://docs.ray.io/en/latest/cluster/running-applications/job-submission/ray-client.html

I have two questions:
1. Can Ray detect and utilize the GPU on macOS?
2. The main issue: When my tasks run on Ubuntu (or Windows for testing), everything works fine and the best hyperparameters are found. However, when any trial is distributed to a macOS device, it enters an infinite loop and does not proceed. (Only CPU, no GPU)

**When trails on Ubuntu device:**
TunerInternal pid=1083113) ╭────────────────────────────────────────────────────────────╮
(TunerInternal pid=1083113) │ Trial train_mnist_81cb3_00004 result                       │
(TunerInternal pid=1083113) ├────────────────────────────────────────────────────────────┤
(TunerInternal pid=1083113) │ checkpoint_dir_name                      checkpoint_000009 │
(TunerInternal pid=1083113) │ time_this_iter_s                                   1.87477 │
(TunerInternal pid=1083113) │ time_total_s                                      17.68371 │
(TunerInternal pid=1083113) │ training_iteration                                      10 │
(TunerInternal pid=1083113) │ accuracy                                             0.976 │
(TunerInternal pid=1083113) ╰────────────────────────────────────────────────────────────╯
129/938 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.2947 - loss: 2.0630
(train_mnist pid=1485046, ip=172.28.1.206) Checkpoint successfully created at: Checkpoint(filesystem=local, path=/home/wanin/ray_results/train_mnist_2024-06-07_06-45-19/train_mnist_81cb3_00004_4_dropout=0.4813,learning_rate=0.0010,units=128_2024-06-07_06-45-20/checkpoint_000009)
282/938 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.9013 - loss: 0.3622
415/938 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.7937 - loss: 0.6934
218/938 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.3873 - loss: 1.8767
312/938 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.9012 - loss: 0.3613
458/938 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8016 - loss: 0.6677
309/938 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.4511 - loss: 1.7279
330/938 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.9012 - loss: 0.3606
483/938 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8057 - loss: 0.6542
403/938 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.4987 - loss: 1.6075
348/938 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.9013 - loss: 0.3599
508/938 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.8095 - loss: 0.6416
......

**When trails on MacOS device:** 
It will repeat below info on the cmd forever 😢 
(TERMINATED  part are finished by Ubuntu device, RUNNING part are running on MACOS device)

(TunerInternal pid=1083113) Trial status: 17 TERMINATED | 3 RUNNING
(TunerInternal pid=1083113) Current time: 2024-06-07 06:51:51. Total running time: 6min 30s
(TunerInternal pid=1083113) Logical resource usage: 24.0/60 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:G)
(TunerInternal pid=1083113) ╭───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
(TunerInternal pid=1083113) │ Trial name                status         units     dropout     learning_rate     iter     total time (s)     accuracy │
(TunerInternal pid=1083113) ├───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
(TunerInternal pid=1083113) │ train_mnist_81cb3_00003   RUNNING          512    0.308015            1e-05                                           │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00005   RUNNING          512    0.495713            1e-05                                           │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00006   RUNNING          256    0.205733            0.001                                           │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00000   TERMINATED       256    0.338235            0.0001       10           12.3224        0.9657 │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00001   TERMINATED       512    0.46337             1e-05         1            5.58101       0.8046 │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00002   TERMINATED       512    0.300281            0.0001        2            8.90747       0.9372 │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00004   TERMINATED       128    0.481268            0.001        10           17.6837        0.976  │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00007   TERMINATED       256    0.295523            0.0001        1            3.47238       0.9148 │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00008   TERMINATED       512    0.473822            0.0001        2            8.5184        0.9321 │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00009   TERMINATED       256    0.366733            0.001        10           28.4771        0.9808 │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00010   TERMINATED       256    0.43327             0.0001        1            2.08663       0.9114 │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00011   TERMINATED       512    0.482333            0.0001        2            9.281         0.9334 │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00012   TERMINATED       256    0.469737            0.0001        1            2.06703       0.9099 │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00013   TERMINATED       256    0.372475            0.001        10           28.37          0.9814 │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00014   TERMINATED       512    0.232035            1e-05         1            2.81465       0.8246 │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00015   TERMINATED       256    0.373846            0.001        10           12.2804        0.9815 │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00016   TERMINATED       256    0.204221            0.001        10           26.9445        0.9811 │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00017   TERMINATED       256    0.421083            1e-05         1            4.26744       0.7503 │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00018   TERMINATED       256    0.340412            0.001         8           10.0374        0.9789 │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00019   TERMINATED       256    0.46581             0.0001        1            3.45041       0.9052 │
(TunerInternal pid=1083113) ╰───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
(TunerInternal pid=1083113) Trial status: 17 TERMINATED | 3 RUNNING
(TunerInternal pid=1083113) Current time: 2024-06-07 06:52:21. Total running time: 7min 0s
(TunerInternal pid=1083113) Logical resource usage: 24.0/60 CPUs, 0/1 GPUs (0.0/1.0 accelerator_type:G)
(TunerInternal pid=1083113) ╭───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
(TunerInternal pid=1083113) │ Trial name                status         units     dropout     learning_rate     iter     total time (s)     accuracy │
(TunerInternal pid=1083113) ├───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
(TunerInternal pid=1083113) │ train_mnist_81cb3_00003   RUNNING          512    0.308015            1e-05                                           │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00005   RUNNING          512    0.495713            1e-05                                           │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00006   RUNNING          256    0.205733            0.001                                           │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00000   TERMINATED       256    0.338235            0.0001       10           12.3224        0.9657 │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00001   TERMINATED       512    0.46337             1e-05         1            5.58101       0.8046 │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00002   TERMINATED       512    0.300281            0.0001        2            8.90747       0.9372 │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00004   TERMINATED       128    0.481268            0.001        10           17.6837        0.976  │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00007   TERMINATED       256    0.295523            0.0001        1            3.47238       0.9148 │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00008   TERMINATED       512    0.473822            0.0001        2            8.5184        0.9321 │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00009   TERMINATED       256    0.366733            0.001        10           28.4771        0.9808 │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00010   TERMINATED       256    0.43327             0.0001        1            2.08663       0.9114 │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00011   TERMINATED       512    0.482333            0.0001        2            9.281         0.9334 │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00012   TERMINATED       256    0.469737            0.0001        1            2.06703       0.9099 │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00013   TERMINATED       256    0.372475            0.001        10           28.37          0.9814 │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00014   TERMINATED       512    0.232035            1e-05         1            2.81465       0.8246 │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00015   TERMINATED       256    0.373846            0.001        10           12.2804        0.9815 │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00016   TERMINATED       256    0.204221            0.001        10           26.9445        0.9811 │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00017   TERMINATED       256    0.421083            1e-05         1            4.26744       0.7503 │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00018   TERMINATED       256    0.340412            0.001         8           10.0374        0.9789 │
(TunerInternal pid=1083113) │ train_mnist_81cb3_00019   TERMINATED       256    0.46581             0.0001        1            3.45041       0.9052 │
(TunerInternal pid=1083113) ╰───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

### Versions / Dependencies

- Head node: Ubuntu with 12 CPUs and 1 GPU
- Worker nodes: One Ubuntu with 24 CPUs, one Mac Studio M2, and one Mac Mini M1.
- python version: 3.10.14
- ray version: 2.20.0
- tf version: 2.16.1

### Reproduction script

Here are my testing code:
```python
import tensorflow as tf
from tensorflow.keras import layers, datasets, models
import ray
from ray import train, tune
from ray.tune.schedulers import ASHAScheduler
import time
from ray.air.integrations.keras import ReportCheckpointCallback

ray.init("ray://localhost:10001")

def build_model(config):
    model = models.Sequential()
    model.add(layers.Flatten(input_shape=(28, 28)))
    model.add(layers.Dense(units=config['units'], activation='relu'))
    model.add(layers.Dropout(rate=config['dropout']))
    model.add(layers.Dense(10, activation='softmax'))

    model.compile(
        optimizer=tf.keras.optimizers.Adam(config['learning_rate']),
        loss='sparse_categorical_crossentropy',
        metrics=['accuracy']
    )
    return model

def train_mnist(config):
    print("Starting training with config:", config)
    start_time = time.time()
    
    physical_devices = tf.config.list_physical_devices('GPU')
    if len(physical_devices) > 0:
        tf.config.experimental.set_memory_growth(physical_devices[0], True)
        print("TensorFlow is using GPU:", physical_devices[0])
    else:
        print("No GPU available, using CPU.")

    (x_train, y_train), (x_test, y_test) = datasets.mnist.load_data()
    x_train, x_test = x_train / 255.0, x_test / 255.0

    with tf.device('/GPU:0'):
        model = build_model(config)

    history = model.fit(
        x_train, y_train,
        validation_data=(x_test, y_test),
        epochs=10,  
        batch_size=64,
        callbacks=[ReportCheckpointCallback(metrics={"accuracy": "val_accuracy"})]
    )
    val_acc = history.history['val_accuracy'][0]
    print("Training completed in", time.time() - start_time, "seconds")
    tune.report(accuracy=val_acc)

search_space = {
    "units": tune.choice([128, 256, 512]),  
    "dropout": tune.uniform(0.2, 0.5), 
    "learning_rate": tune.choice([1e-3, 1e-4, 1e-5]) 
}

tuner = tune.Tuner(
    tune.with_resources(
        tune.with_parameters(train_mnist),
        resources={"cpu": 8}
    ),
    # tune.with_parameters(train_mnist),
    param_space=search_space,
    tune_config=tune.TuneConfig(
        scheduler=scheduler,
        num_samples=20
    ),
    run_config=train.RunConfig(
            storage_path="nfs path",
        )
)
results = tuner.fit()
```

### Issue Severity

None
